sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/perPartitionAutomaticFailover/GlobalPartitionEndpointManagerForPerPartitionAutomaticFailover.java

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.perPartitionAutomaticFailover;
+
+import com.azure.cosmos.implementation.GlobalEndpointManager;
+import com.azure.cosmos.implementation.OperationType;
+import com.azure.cosmos.implementation.PartitionKeyRange;
+import com.azure.cosmos.implementation.PartitionKeyRangeWrapper;
+import com.azure.cosmos.implementation.ResourceType;
+import com.azure.cosmos.implementation.RxDocumentServiceRequest;
+import com.azure.cosmos.implementation.apachecommons.lang.StringUtils;
+import com.azure.cosmos.implementation.routing.RegionalRoutingContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class GlobalPartitionEndpointManagerForPerPartitionAutomaticFailover {
+    private static final Logger logger = LoggerFactory.getLogger(GlobalPartitionEndpointManagerForPerPartitionAutomaticFailover.class);
+    private final ConcurrentHashMap<PartitionKeyRangeWrapper, PartitionLevelFailoverInfo> partitionKeyRangeToLocation;
+    private final GlobalEndpointManager globalEndpointManager;
+    private final boolean isPerPartitionAutomaticFailoverEnabled;
+
+    public GlobalPartitionEndpointManagerForPerPartitionAutomaticFailover(
+        GlobalEndpointManager globalEndpointManager,
+        boolean isPerPartitionAutomaticFailoverEnabled) {
+
+        this.globalEndpointManager = globalEndpointManager;
+        this.partitionKeyRangeToLocation = new ConcurrentHashMap<>();
+        this.isPerPartitionAutomaticFailoverEnabled = isPerPartitionAutomaticFailoverEnabled;
+    }
+
+    public boolean tryAddPartitionLevelLocationOverride(RxDocumentServiceRequest request) {
+
+        if (!this.isPerPartitionAutomaticFailoverEnabled) {
+            return false;
+        }
+
+        checkNotNull(request, "Argument 'request' cannot be null!");
+        checkNotNull(request.requestContext, "Argument 'request.requestContext' cannot be null!");
+
+
+        if (request.getResourceType() != ResourceType.Document) {
+            return false;
+        }
+
+        if (request.getOperationType() == OperationType.QueryPlan) {
```

Comment:
what about address refresh operations?

Reply:
Address refresh requests are of type `HttpRequest`. This class deals with `RxDocumentServiceRequest` which points to non-address requests, so no need of special handling here.